### ISSUE TITLE
Issue when using where() to filter causing extra reads

### ISSUE THREAD

**Issue by momoDragon** [DarshanGowda0/GeoFlutterFire#29]
**Code**
`collectionRef: Firestore.instance
                .collection('vendors')
                .where('storeType', isEqualTo: 'Cafe)
                .where('vendorType', isEqualTo: 2).where('block', isEqualTo: false)
                .where('applicationStatus', isEqualTo: 3).where('isVisible', isEqualTo: true)`

**Scenario**
As u can see i am using where(). However its causing me extra reads.
I have 100 items in 10km radius, but only one of them meets the criteria 
`.where('storeType', isEqualTo: 'Cafe)`
But this caused me 100 reads.

**Comment by japborst** [comment]
I believe each `.where()` is basically an `OR` statement that is being merged when they arrive, so it kinda makes sense that it results in multiple reads.

**Comment by momoDragon** [comment]
Not possible to incorporate .where() when doing the geoquery rather than do it after the query? Because without its, its not very flexible for doing querys  hence causing extra unnecessary doc reads

**Comment by DarshanGowda0** [comment]
@momoDragon `.where()` is supposed to just give you 1 read, I'm sure this has been the case from the first version when used with `isEqualTo`, it uses the same query as cloud_firestore. Could you verify again?

**Comment by mikeyyg96** [comment]
I'm surprised you got a read at all. I tried filtering my collection with a where clause and it seemed to have not work. I checked multiple times to see if it is right collection, field, and value and they were all correct, but still no read. When I would take the where clause out, all the records come through.

**Comment by DarshanGowda0** [comment]
@mikeyyg96 `.where()` works only with `isEqualTo` at the moment. And also are you creating the index the first time?

**Comment by mikeyyg96** [comment]
@DarshanGowda0 Yeah I used the exact format in the documentation:
`var queryRef = _firestore.collection('locations').where('city', isEqualTo: 'bangalore');`
but with my values:
`var queryRef = _firestore.collection('events').where('category', isEqualTo: 'activities');`
I have this field in my collection and that value as well but nothing returns. When I take out the where clause, the record pops up. It is only 1 record in my collection.

**Comment by momoDragon** [comment]
Yes i have tested. and ya. its quite bad. its causing me hell alot of reads man. @DarshanGowda0 u need to try it again man. Open a new project and use app engine to track the usage.

**Comment by momoDragon** [comment]
> @DarshanGowda0 Yeah I used the exact format in the documentation:
> `var queryRef = _firestore.collection('locations').where('city', isEqualTo: 'bangalore');`
> but with my values:
> `var queryRef = _firestore.collection('events').where('category', isEqualTo: 'activities');`
> I have this field in my collection and that value as well but nothing returns. When I take out the where clause, the record pops up. It is only 1 record in my collection.
> ![Screenshot_20190501-153146__01](https://user-images.githubusercontent.com/33616962/57038103-17c7b880-6c27-11e9-960c-9b8a7e19bc54.jpg)

Check your console. it would prompt u to create index. there will be a link. just click it

**Comment by mikeyyg96** [comment]
@momoDragon Can you please show me where that link is located or what exactly is it saying? I have a lot of information in my console and it keeps updating and filtering the console with vscode sucks.

**Comment by momoDragon** [comment]
@mikeyyg96  yup. I always use strict mode

**Comment by DarshanGowda0** [comment]
Hey @momoDragon! I verified this again, the number of reads is always 1 if I have just one document that satisfies the `where()` condition. The firestore query you pass is the first condition that is executed and geofire filtering happens after the `where()` filter is done (check [this](https://github.com/DarshanGowda0/GeoFlutterFire/blob/ca14deb1cf899d4fed9011dbd447a8e1323e312a/lib/src/collection.dart#L145)).

**Comment by rmelian** [comment]
I have sort of a similar issue. When I use a where clause the stream never returns back:
```dart
var query = Firestore.instance
        .collection('itineraries')
        .where('status', isEqualTo: status.toString());

    GeoFirePoint center =
        geo.point(latitude: nearby.latitude, longitude: nearby.longitude);

    double radius = 50;
    String field = 'position';

    Stream<List<DocumentSnapshot>> stream = geo
        .collection(collectionRef: query)
        .within(center: center, radius: radius, field: field);

    List<DocumentSnapshot> docs = await stream.first;
```
if I remove the where it works fine. I have created the composite index in different ways manually as I never get the link to create it
![image](https://user-images.githubusercontent.com/4565267/59558713-57274a00-8fc6-11e9-8446-1415baee3ea4.png)
![image](https://user-images.githubusercontent.com/4565267/59558716-6dcda100-8fc6-11e9-81ef-d125cca34fba.png)
 Am I missing something here?

**Comment by Moellendorff** [comment]
Same issue as rmelian's. Like rmelian, I never get the link to create the composite index.

**Comment by Moellendorff** [comment]
@rmelian Solved by manually creating an index that, in your terms, would be:
Collection ID: itineraries
Field indexed: status Ascending position.geohash Ascending
Query scope: Collection

**Comment by yevgeniaronov** [comment]
i have sort of the same issue:
`    var collectionRef = Firestore.instance.collection('/user_data').where('gender', isEqualTo: 'male');
    return geo
        .collection(collectionRef: collectionRef.reference())
        .within(center: center, radius: 1, field: 'location');`

it totally ignores the .where, and just outputs all users within the radius.

**Comment by atozbtcom** [comment]
> @rmelian Solved by manually creating an index that, in your terms, would be:
> Collection ID: itineraries
> Field indexed: status Ascending position.geohash Ascending
> Query scope: Collection

is need sequins as where conditions and then position.geohash index look like status_position.geohash

**Comment by momoDragon** [comment]
@atozbtcom  and @Moellendorff are right. I just verified that. 
@yevgeniaronov  u need to manually set indexing in firebase console.

ensure that u sort the where clause field first before the position.geohash
<img width="895" alt="Screenshot 2020-01-01 at 11 27 08 AM" src="https://user-images.githubusercontent.com/40727575/71637844-b0687980-2c89-11ea-8525-da1846bd98f9.png">

**Comment by atnegrete** [comment]
@momoDragon So position.geohash must be the last field when setting up the index?

**Comment by smolugu** [comment]
query still not working after creating composite index. did anyone find a solution for this? Thank you!
